docs: graceful token refresh (#1938)
ory/docs: docs/hydra/guides/graceful-token-refresh.mdx

---
id: graceful-token-refresh
title: Graceful token refresh
---

# Graceful refresh token rotation in Ory OAuth2 / Ory Hydra

Graceful refresh token rotation is a feature in Ory OAuth2 / Ory Hydra that allows for a smoother transition during refresh token
usage. With this feature enabled, a refresh token remains valid within a defined grace period, allowing multiple usages without
immediate invalidation. This can be beneficial in scenarios where network issues or delayed token exchanges may otherwise disrupt
session continuity.

When enabled, using a refresh token marks it as "used" in the database and extends its expiration time by the duration of the
configured grace period. As long as the grace period is active, subsequent token refreshes will return new access and refresh
tokens. All tokens issued in this grace period remain linked in a single token chain, so revoking one refresh token or consent
will invalidate all associated tokens.

> ## Enabling graceful refresh token rotation

Reviewer: ```suggestion
## Enable graceful refresh token rotation
```

To enable graceful refresh token rotation with a specific grace period (for example, 60 seconds), run the following command:

```shell
ory patch oauth2-config --project <project-id> --workspace <workspace-id> \
  --replace "/oauth2/grant/refresh_token/rotation_grace_period=60s"
```

In this command:

- `<project-id>` and `<workspace-id>` should be replaced with your specific project and workspace identifiers.
- The `rotation_grace_period` specifies the grace period duration. Here, `60s` sets a 60-second grace period.

> ## Disabling graceful refresh token rotation

Reviewer: ```suggestion
## Disable graceful refresh token rotation
```

To disable the graceful refresh token rotation, remove the `rotation_grace_period` parameter using the command below:

```shell
ory patch oauth2-config --project <project-id> --workspace <workspace-id> \
  --remove "/oauth2/grant/refresh_token/rotation_grace_period"
```

> ## Configuration in self-hosted deployments (Ory Enterprise License / DIY)
> 
> For self-hosted deployments, you can configure graceful refresh token rotation in your configuration file:

Reviewer: ```suggestion
## Configuration in self-hosted deployments 

For self-hosted deployments with Ory Enterprise License or Ory OSS, you can configure graceful refresh token rotation in your configuration file:
```

If possible no parentheses in headings 🙏
